testsuites/syncgateway/functional/test_sg_replicate.py

```diff
@@ -0,0 +1,259 @@
+    chans = sg1A_user.channels
+    sg1.start_push_replication(sg2.admin_url, DB1, DB2, continuous=False, use_remote_source=True, channels=chans)
+
+    # Verify that the doc added to sg1 made it to sg2
+    assert_has_doc(sg2_user, doc_id_sg1A)
+
+
+def test_sg_replicate_continuous_replication():
+
+    sg1, sg2 = create_sync_gateways()
+
+    # Create users (in order to add docs)
+    sg1_user, sg2_user = create_sg_users(sg1, sg2, DB1, DB2)
+
+    # Kick off continuous replication
+    sg1.start_push_replication(sg2.admin_url, DB1, DB2, continuous=True, use_remote_source=True)
+
+    # Add docs
+    doc_id = sg1_user.add_doc()
+
+    # Wait til all docs sync to target
+    wait_until_docs_sync(sg2_user, [doc_id])
+
+    # Shutdown target
+    sg2.stop()
+
+    # Add more docs
+    doc_id_2 = sg1_user.add_doc()
+
+    # Wait a few seconds to give the source replicator time to have some attempts
+    time.sleep(5)
+
+    # Restart target
+    sg2.start(config=CONFIG_PATH)
+
+    # Wait til all docs sync to target
+    wait_until_docs_sync(sg2_user, [doc_id, doc_id_2])
+
+    # Stop replications
+    sg1.stop_push_replication(sg2.admin_url, DB1, DB2, continuous=True, use_remote_source=True)
+
+    # Wait until active_tasks is empty (or throw exception)
+    wait_until_active_tasks_empty(sg1)
+
+    # Add more docs, even though the replication is already stopped
+    doc_id_3 = sg1_user.add_doc()
+
+    # Wait a few seconds to give it time to potentially propagate
+    time.sleep(5)
+
+    # Make sure the doc did not propagate to the target
+    assert_does_not_have_doc(sg2_user, doc_id_3)
+
+
+def test_sg_replicate_delete_db_replication_in_progress():
+
+    sg1, sg2 = create_sync_gateways()
+
+    # Kick off continuous replication
+    sg1.start_push_replication(sg2.admin_url, DB1, DB2, continuous=True, use_remote_source=True)
+
+    # Wait until active_tasks is non empty
+    wait_until_active_tasks_non_empty(sg1)
+
+    # Delete the database
+    sg1.delete_db(DB1)
+    sg2.delete_db(DB2)
+
+    # Query active tasks and make sure the replication is gone
+    wait_until_active_tasks_empty(sg1)
+
+
+def test_sg_replicate_non_existent_db():
+
+    sg1, sg2 = create_sync_gateways()
+
+    # Start a push replication
+    got_exception = False
+    try:
+        sg1.start_push_replication(sg2.admin_url, DB1, DB2, continuous=False, use_remote_source=True)
+    except:
+        got_exception = True
+
+    assert got_exception == True
+
+
+def test_sg_replicate_push_async():
+
+    # if the async stuff works, we should be able to kick off a large
+    # push replication and get a missing doc before the replication has
+    # a chance to finish.  And then we should later see that doc.
+
+    sg1, sg2 = create_sync_gateways()
+
+    admin = Admin(sg1)
+    admin.admin_url = sg1.url
+
+    sg1_user, sg2_user = create_sg_users(sg1, sg2, DB1, DB2)
+
+    # Add docs to sg1
+    doc_ids_added = []
+    last_doc_id_added = None
+    for i in xrange(100):
```

Comment:
Do you mind deferring that parameterization until we're actually calling this method with different values?